python-avd/pyavd/_eos_designs/structured_config/network_services/utils.py

```diff
@@ -402,3 +402,54 @@ def get_vlan_aware_bundle_rt(
             return f"{admin_subfield}:{rt_override}"
 
         return f"{admin_subfield}:{bundle_number}"
+
+    def get_vrf_router_id(
+        self: AvdStructuredConfigNetworkServices,
+        vrf: EosDesigns._DynamicKeys.DynamicNetworkServicesItem.NetworkServicesItem.VrfsItem,
+        router_id: str,
+        tenant_name: str,
+    ) -> str | None:
+        """
+        Determine the router ID for a given VRF based on its configuration.
+
+        Args:
+            vrf (VrfsItem): The VRF object containing OSPF/BGP and vtep_diagnostic details.
+            router_id (str): The router ID type specified for the VRF (e.g., "vtep_diagnostic", "main_router_id", "none", or "IPv4 address").
+            tenant_name (str): The name of the tenant to which the VRF belongs.
+
+        Returns:
+            str | None: The resolved router ID as a string, or None if the router ID is not applicable.
+
+        Raises:
+            AristaAvdInvalidInputsError: If required configuration for "vtep_diagnostic" router ID is missing.
+        """
+        # Handle "vtep_diagnostic" router ID case
+        if router_id == "vtep_diagnostic":
+            vrf_diagnostic_loopback = vrf.vtep_diagnostic.loopback
+            loopback_ip_pools = vrf.vtep_diagnostic.loopback_ip_pools
+
+            # Validate required configuration
+            if vrf_diagnostic_loopback is None or loopback_ip_pools is None:
+                msg = (
+                    f"Invalid configuration on VRF '{vrf.name}' in Tenant '{tenant_name}'. "
+                    "Both 'vtep_diagnostic.loopback' and 'vtep_diagnostic.loopback_ip_pools' must be defined "
+                    "when 'router_id' is set to 'vtep_diagnostic' on the VRF."
+                )
+                raise AristaAvdInvalidInputsError(msg)
+
+            # Resolve router ID from loopback interface
+            interface_data = self._get_vtep_diagnostic_loopback_for_vrf(vrf)
+            return interface_data["ip_address"].split("/")[0]
```

Comment:
there is a method to get ip address from prefix. It can be used here.

Reply:
updated